domain/machine/state/reboot.go

```diff
@@ -87,3 +106,112 @@ func (st *State) IsMachineRebootRequired(ctx context.Context, uuid string) (bool
 
 	return isRebootRequired, errors.Annotatef(err, "requiring reboot of machine %q", uuid)
 }
+
+// GetParentMachine fetches the UUID of the parent machine based on the provided UUID.
+// It returns the UUID of the parent machine if found, otherwise returns an errors.NotFound.
+//
+// The function uses a database connection from the State struct to execute the query.
+// The `ctx` parameter is the context.Context used for the transaction.
+// The `uuid` parameter is the UUID of the machine to get the parent machine for.
+//
+// If the database query returns no rows, the function returns a errors.ParentNotFound error
+// Any other errors that occur during the execution of the function are annotated with the UUID of the machine being fetched.
+func (st *State) GetParentMachine(ctx context.Context, uuid string) (string, error) {
+	db, err := st.DB()
+	if err != nil {
+		return "", errors.Trace(err)
+	}
+
+	var parentMachine machineUUID
+	getParentQuery := `SELECT machine_parent.parent_uuid as &machineUUID.uuid  FROM machine_parent WHERE machine_uuid = $machineUUID.uuid`
+	getParentStmt, err := sqlair.Prepare(getParentQuery, machineUUID{})
+	if err != nil {
+		return "", errors.Trace(err)
+	}
+	err = db.Txn(ctx, func(ctx context.Context, tx *sqlair.TX) error {
+		err := tx.Query(ctx, getParentStmt, machineUUID{uuid}).Get(&parentMachine)
+		return errors.Trace(err)
+	})
+	if errors.Is(err, sqlair.ErrNoRows) {
+		err = machineerrors.ParentNotFound
+	}
+	return parentMachine.UUID, errors.Annotatef(err, "fetching parent machine %q", uuid)
+}
+
+// ShouldRebootOrShutdown determines if a machine should reboot or shutdown based on its state and parent's state.
+//
+// The function first checks if the parent machine requires a reboot. If so, it returns ShouldShutdown immediately.
+//
+// If the parent machine does not require a reboot, the function checks if the current machine requires a reboot. If
+// so, it returns ShouldReboot. If neither the parent machine nor the current machine require a reboot, it returns
+// ShouldDoNothing.
+//
+// The function uses the State.GetParentMachine method to get the parent UUID of a machine. If the parent machine does
+// not exist, the function returns ShouldDoNothing. If there is a grandparent machine, which is not supported, the
+// function returns an errors.GrandParentNotSupported.
+//
+// The State.IsMachineRebootRequired method is used to check if a machine requires a reboot.
+//
+// The function returns an error if there is a problem with the database or if an error occurs during the transaction.
+func (st *State) ShouldRebootOrShutdown(ctx context.Context, uuid string) (RebootAction, error) {
+	db, err := st.DB()
+	if err != nil {
+		return ShouldDoNothing, errors.Trace(err)
+	}
+
+	var parentShouldReboot, machineShouldReboot bool
+	err = db.Txn(ctx, func(ctx context.Context, tx *sqlair.TX) error {
+		// Get parent UUID
+		parentUUID, err := st.GetParentMachine(ctx, uuid)
```

Comment:
**issue (blocker):** Don't nest transactions. Just copy and paste the statements.